tests/cpp/test_pointwise.cpp

```diff
@@ -1435,4 +1437,206 @@ TEST_F(
+  // test [dim0], [dim0, 2], [dim0, 2, 4]
+  std::vector<int64_t> element_at_each_dim(ndims);
+  element_at_each_dim[0] = dim0;
+  int64_t total_elem_count = dim0;
+  for (int64_t i = 1; i < ndims; i++) {
+    int64_t dim_i = 1 << i;
+    element_at_each_dim[i] = dim_i;
+    total_elem_count *= dim_i;
+  }
+  if (total_elem_count * dtype_bytes % 16 != 0) {
+    GTEST_SKIP() << "Total bytes is not divisible by 16, can't use TMA, "
+                    "total_elem_count: "
+                 << total_elem_count << ", dtype_bytes: " << dtype_bytes;
+    return;
+  }
+
+  auto fusion_ptr = std::make_unique<Fusion>();
+  FusionGuard fg(fusion_ptr.get());
+  Fusion& fusion = *fusion_ptr;
+  auto tv0 = makeContigTensor(ndims);
+  fusion.addInput(tv0);
+  auto tv1 = add(tv0, tv0);
+  fusion.addOutput(tv1);
+
+  // Create TMA loads from inputs to shared memory
+  auto tv0_smem = tv0->cacheAfter(LoadStoreOpType::CpAsyncBulkTensorTile);
+  tv0_smem->setMemoryType(MemoryType::Shared);
+
+  // Cache loads from shared memory to registers
+  auto tv0_regs = tv0_smem->cacheAfter();
+
+  // Output caching: regs -> [smem ->] global memory
+  TensorView* tv1_smem = nullptr;
+  TensorView* tv1_regs = nullptr;
+  if (use_tma_store) {
+    // TMA store path: regs -> smem -> global memory (via TMA)
+    tv1_smem = tv1->cacheBefore(LoadStoreOpType::CpAsyncBulkTensorTile);
+    tv1_smem->setMemoryType(MemoryType::Shared);
+    tv1_regs = tv1_smem->cacheBefore();
+  } else {
+    // Regular store path: regs -> global memory (no TMA)
+    tv1_regs = tv1->cacheBefore();
+  }
+
+  TensorView* reference = tv1;
+
+  // Step 1: Create TMA domain by merging all domains into one, then split into
+  // 2 TMA domains. Requirements: all domains must be contiguous and the TMA
+  // domain split must be evenly divisible.
+  // Transformation: [I0, I1, ...] -> [ALL_DIMS] -> [D0, D1]
+  for (int64_t i = 0; i < ndims - 1; i++) {
+    reference->merge(0);
+  }
+
+  int64_t D1 = scheduler_utils::getInnerTmaDomainSize(
+      total_elem_count, 512, dtype_bytes);
+  int64_t D0 = total_elem_count / D1;
+  NVF_ERROR(
+      total_elem_count % D1 == 0,
+      "TMA domain can only be created with divisible split, D1: ",
+      D1,
+      " total_elem_count: ",
+      total_elem_count);
+  reference->split(0, D1);
+
+  // Step 2: Define 2D box/tile (using dense tile where box equals tile).
+  // Select an IterDomain in the TMA domain, then split that IterDomain by the
+  // box size for that dimension. Requirement: D1/ti > 1, otherwise to and ti
+  // would be considered as a single box dimension.
+  // Transformation: [D0, D1] -> [D0/to, to, D1/ti, ti]
+  int64_t tma_tile_size = 4096;
+  int64_t ti = std::min(256L, std::max(1L, D1 / 2)),
+          to = std::min(256L, std::min(tma_tile_size / ti, D0));
+  reference->split(0, to);
+  reference->split(2, ti);
+
+  // Check grid dimensions and swap parallel types if needed to avoid exceeding
+  // the maximum grid dimension limit (65535).
```

Comment:
I was not aware BIDx had a different limit than BIDy and BIDz. Perhaps modify the comment to be more clear about this:
`... to avoid exceeding the maximum BIDy dimension limit (65535).`

Alternatively, why not always assign `ceilDiv(D0, to)` to BIDx so that we don't need this logic. Is it slower to do so?

Reply:
Yes, perf is slightly lower, there is a figure in the design doc showing perf drop due to swap. Heuristic should avoid swapping.